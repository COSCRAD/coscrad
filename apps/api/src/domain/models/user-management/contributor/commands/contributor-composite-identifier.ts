import { AggregateType } from '@coscrad/api-interfaces';
import { UUID } from '@coscrad/data-types';
import { AggregateTypeProperty } from '../../../shared/common-commands';

export class ContributorCompositeIdentifier {
    @AggregateTypeProperty([AggregateType.contributor])
    type = AggregateType.contributor;

    @UUID({
        label: 'ID',
        description: 'unique identifier',
    })
    id: string;
}
