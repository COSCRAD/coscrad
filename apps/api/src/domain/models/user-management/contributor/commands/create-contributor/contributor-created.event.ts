import { CoscradEvent } from '../../../../../../domain/common';
import { BaseEvent } from '../../../../shared/events/base-event.entity';
import { CreateContributor } from './create-contributor.command';

export type ContributorCreatedPayload = CreateContributor;

@CoscradEvent('CONTRIBUTOR_CREATED')
export class ContributorCreated extends BaseEvent {
    readonly type = 'CONTRIBUTOR_CREATED';
}
