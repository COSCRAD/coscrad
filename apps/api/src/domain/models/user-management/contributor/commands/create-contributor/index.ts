export * from './contributor-created.event';
export * from './create-contributor.command';
export * from './create-contributor.command-handler';
