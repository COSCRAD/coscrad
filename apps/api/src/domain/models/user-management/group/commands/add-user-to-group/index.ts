export * from './add-user-to-group.command';
export * from './add-user-to-group.command-handler';
