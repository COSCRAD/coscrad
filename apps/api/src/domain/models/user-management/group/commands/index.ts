export * from './add-user-to-group';
export * from './create-group';
