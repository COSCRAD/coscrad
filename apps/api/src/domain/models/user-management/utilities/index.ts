export * from './coscrad-date.entity';
export * from './invalid-date.error';
