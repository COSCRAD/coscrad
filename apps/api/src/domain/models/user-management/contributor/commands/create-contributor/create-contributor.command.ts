import { AggregateCompositeIdentifier, ICommandBase } from '@coscrad/api-interfaces';
import { Command } from '@coscrad/commands';
import { NestedDataType, NonEmptyString } from '@coscrad/data-types';
import { ContributorCompositeIdentifier } from '../contributor-composite-identifier';

@Command({
    type: 'CREATE_CONTRIBUTOR',
    label: 'Create Contributor',
    description: 'Create a new COSCRAD contributor',
})
export class CreateContributor implements ICommandBase {
    @NestedDataType(ContributorCompositeIdentifier, {
        label: 'Composite Identifier',
        description: 'system-wide unique identifier',
    })
    readonly aggregateCompositeIdentifier: AggregateCompositeIdentifier;

    @NonEmptyString({
        label: "the contributor's first name",
        description: "The Contributor's first name",
    })
    readonly firstName: string;

    @NonEmptyString({
        label: "the contributor's last name",
        description: "The Contributor's last name",
    })
    readonly lastName: string;

    @NonEmptyString({
        label: 'short bio',
        description: 'a brief biography to help uniquely identify this contributor',
        isOptional: true,
    })
    readonly shortBio?: string;

    @NonEmptyString({
        label: 'date of birth',
        description: "The Contributor's Date of Birth (YYYY-MM-DD)",
        isOptional: true,
    })
    readonly dateOfBirth?: string;
}
