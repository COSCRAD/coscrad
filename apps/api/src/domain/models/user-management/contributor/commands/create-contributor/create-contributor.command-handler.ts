import { AggregateType } from '@coscrad/api-interfaces';
import { CommandHandler, ICommand } from '@coscrad/commands';
import { isNonEmptyString } from '@coscrad/validation-constraints';
import { Valid } from '../../../../../../domain/domainModelValidators/Valid';
import buildInstanceFactory from '../../../../../../domain/factories/utilities/buildInstanceFactory';
import { DeluxeInMemoryStore } from '../../../../../../domain/types/DeluxeInMemoryStore';
import { InMemorySnapshot } from '../../../../../../domain/types/ResourceType';
import { InternalError, isInternalError } from '../../../../../../lib/errors/InternalError';
import cloneToPlainObject from '../../../../../../lib/utilities/cloneToPlainObject';
import { DTO } from '../../../../../../types/DTO';
import { ResultOrError } from '../../../../../../types/ResultOrError';
import { BaseCreateCommandHandler } from '../../../../shared/command-handlers/base-create-command-handler';
import { BaseEvent, IEventPayload } from '../../../../shared/events/base-event.entity';
import { EventRecordMetadata } from '../../../../shared/events/types/EventRecordMetadata';
import { CoscradDate } from '../../../utilities';
import { CoscradContributor } from '../../entities/coscrad-contributor.entity';
import { ContributorCreated } from './contributor-created.event';
import { CreateContributor } from './create-contributor.command';

@CommandHandler(CreateContributor)
export class CreateContributorCommandHandler extends BaseCreateCommandHandler<CoscradContributor> {
    protected createNewInstance({
        aggregateCompositeIdentifier: { id },
        firstName,
        lastName,
        dateOfBirth,
        shortBio,
    }: CreateContributor): ResultOrError<CoscradContributor> {
        const dateBuildResult = isNonEmptyString(dateOfBirth)
            ? CoscradDate.parseString(dateOfBirth)
            : undefined;

        if (isInternalError(dateBuildResult)) {
            return dateBuildResult;
        }

        const createDto: DTO<CoscradContributor> = {
            type: AggregateType.contributor,
            id,
            fullName: { firstName, lastName },
            dateOfBirth: dateBuildResult,
            shortBio,
        };

        return buildInstanceFactory(CoscradContributor)(createDto);
    }

    protected fetchRequiredExternalState(_command?: ICommand): Promise<InMemorySnapshot> {
        return Promise.resolve(new DeluxeInMemoryStore({}).fetchFullSnapshotInLegacyFormat());
    }

    protected validateExternalState(
        _state: InMemorySnapshot,
        _instance: CoscradContributor
    ): InternalError | Valid {
        return Valid;
    }

    protected buildEvent(
        commandPayload: CreateContributor,
        eventMeta: EventRecordMetadata
    ): BaseEvent<IEventPayload> {
        const eventPayload = cloneToPlainObject(commandPayload);

        if (isNonEmptyString(eventPayload.dateOfBirth)) {
            // Here we convert the string that came from user input to our internal data format
            eventPayload.dateOfBirth = CoscradDate.parseString(commandPayload.dateOfBirth);
        }

        return new ContributorCreated(eventPayload, eventMeta);
    }
}
