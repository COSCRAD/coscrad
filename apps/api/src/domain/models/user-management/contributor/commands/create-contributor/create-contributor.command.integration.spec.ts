import { AggregateType } from '@coscrad/api-interfaces';
import { CommandHandlerService, FluxStandardAction } from '@coscrad/commands';
import { INestApplication } from '@nestjs/common';
import setUpIntegrationTest from '../../../../../../app/controllers/__tests__/setUpIntegrationTest';
import { IIdManager } from '../../../../../../domain/interfaces/id-manager.interface';
import { AggregateId } from '../../../../../../domain/types/AggregateId';
import buildEmptyInMemorySnapshot from '../../../../../../domain/utilities/buildEmptyInMemorySnapshot';
import { ArangoDatabaseProvider } from '../../../../../../persistence/database/database.provider';
import TestRepositoryProvider from '../../../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import { DTO } from '../../../../../../types/DTO';
import { assertCreateCommandError } from '../../../../__tests__/command-helpers/assert-create-command-error';
import { assertCreateCommandSuccess } from '../../../../__tests__/command-helpers/assert-create-command-success';
import { DummyCommandFsaFactory } from '../../../../__tests__/command-helpers/dummy-command-fsa-factory';
import { CommandAssertionDependencies } from '../../../../__tests__/command-helpers/types/CommandAssertionDependencies';
import { buildFakeTimersConfig } from '../../../../__tests__/utilities/buildFakeTimersConfig';
import { dummySystemUserId } from '../../../../__tests__/utilities/dummySystemUserId';
import { Month } from '../../../utilities';
import { CoscradContributor } from '../../entities/coscrad-contributor.entity';
import { CreateContributor } from './create-contributor.command';

const commandType = 'CREATE_CONTRIBUTOR';

const firstName = "the contributor's first name";

const lastName = "the contributor's last name";

const shortBio = "all about the contributor's life";

const dateOfBirth = '1949-05-12';

const expectedDay = 12;

const expectedMonth = Month.May;

const expectedYear = 1949;

const buildValidCommandFSA = (id: AggregateId): FluxStandardAction<DTO<CreateContributor>> => ({
    type: commandType,

    payload: {
        aggregateCompositeIdentifier: { id, type: AggregateType.contributor },
        firstName,
        lastName,
        shortBio,
        dateOfBirth,
    },
});

const initialState = buildEmptyInMemorySnapshot();

const fsaFactory = new DummyCommandFsaFactory<CreateContributor>(buildValidCommandFSA);

describe('CreateContributor', () => {
    let testRepositoryProvider: TestRepositoryProvider;

    let commandHandlerService: CommandHandlerService;

    let app: INestApplication;

    let databaseProvider: ArangoDatabaseProvider;

    let idManager: IIdManager;

    let commandAssertionDependencies: CommandAssertionDependencies;

    beforeAll(async () => {
        ({ testRepositoryProvider, commandHandlerService, idManager, app, databaseProvider } =
            await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
                },
                { shouldMockIdGenerator: true }
            ));

        commandAssertionDependencies = {
            testRepositoryProvider,
            idManager,
            commandHandlerService,
        };

        jest.useFakeTimers(buildFakeTimersConfig());
    });

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();
    });

    afterEach(async () => {
        await testRepositoryProvider.testTeardown();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    const seedValidInitialState = async () => {
        await testRepositoryProvider.addFullSnapshot(initialState);
    };

    describe(`when the command is valid`, () => {
        it('should succeed', async () => {
            await assertCreateCommandSuccess(commandAssertionDependencies, {
                systemUserId: dummySystemUserId,
                buildValidCommandFSA,
                seedInitialState: seedValidInitialState,
                checkStateOnSuccess: async ({
                    aggregateCompositeIdentifier: { id },
                }: CreateContributor) => {
                    const contributorSearchResult = await testRepositoryProvider
                        .getContributorRepository()
                        .fetchById(id);

                    expect(contributorSearchResult).toBeInstanceOf(CoscradContributor);

                    const contributor = contributorSearchResult as CoscradContributor;

                    const fullName = contributor.fullName;

                    expect(fullName.firstName).toBe(firstName);

                    expect(fullName.lastName).toBe(lastName);

                    const foundDob = contributor.dateOfBirth;

                    expect(foundDob.day).toBe(expectedDay);

                    expect(foundDob.month).toBe(expectedMonth);

                    expect(foundDob.year).toBe(expectedYear);

                    expect(contributor.shortBio).toBe(shortBio);
                },
            });
        });
    });

    describe('when the command is invalid', () => {
        describe('when neither short bio nor date of birth is provided', () => {
            it('should return the expected error', async () => {
                await assertCreateCommandError(commandAssertionDependencies, {
                    systemUserId: dummySystemUserId,
                    buildCommandFSA: (id: AggregateId) =>
                        fsaFactory.build(id, {
                            shortBio: undefined,
                            dateOfBirth: undefined,
                        }),
                    seedInitialState: seedValidInitialState,
                });
            });
        });
    });
});
