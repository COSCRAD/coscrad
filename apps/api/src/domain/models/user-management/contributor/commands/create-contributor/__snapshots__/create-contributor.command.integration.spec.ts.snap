// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateContributor when the command is valid should succeed: 41fb2d7f-c483-4e09-a1f0-e9909a6b0001 1`] = `
CoscradContributor {
  "eventHistory": [
    {
      "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0002",
      "meta": {
        "dateCreated": 1664237194356,
        "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0002",
        "userId": "system4d-3b7d-4bad-9bdd-2b0d70userid",
      },
      "payload": {
        "aggregateCompositeIdentifier": {
          "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0001",
          "type": "contributor",
        },
        "firstName": "the contributor's first name",
        "lastName": "the contributor's last name",
      },
      "type": "CONTRIBUTOR_CREATED",
    },
  ],
  "fullName": FullName {
    "firstName": "the contributor's first name",
    "lastName": "the contributor's last name",
  },
  "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0001",
  "type": "contributor",
}
`;
